### ISSUE TITLE
Shift period not generating comma for ZX81

### ISSUE THREAD

**Issue by ikjordan** [charlierobson/EightyOne#43]
When shift period is pressed, the greater than symbol is generated in EightyOne when emulating a ZX81, rather than the comma symbol. 

Normally this would be trivial, but I have recently made possibly the world's worst keyboard by creating a USB keyboard from a ZX81 matrix and a 3d printed ZX81 case. Due to this issue I have no way to generate a comma using the keyboard in EightyOne. I can generate a comma when using sz81, CLK, vb81and ZEsarUX.

I believe the issue is in  kbstatus.cpp: 
The first check for period in the table appears to use VK_DECIMAL (110, 0x6E). I believe that keycode refers to the decimal key on the numerical keypad. When the period key on the main part of the keyboard is pressed VK_OEM_PERIOD (190) is generated. This means (for the ZX81) line 102 of kbstatus.cpp does not generally match, and instead line 116 is used, which explicitly generates greater than for shift period.

I cannot introduce a workaround in my keyboard code, and generate VK_DECIMAL, as windows converts shift VK_DECIMAL to VK_DELETE. Also that would mean the keyboard would not work with the other emulators.

The fix in EightyOne appears to be to replace instances of VK_DECIMAL  with  VK_OEM_PERIOD. I have not installed the development environment to rebuild and check this, but I have edited the EightyOne executable with a hex editor. I replaced the 4 instances (ZX80, ZX81, Lambda, Spectrum) in the keyboard table of VK_DECIMAL (0x6E) with VK_OEM_PERIOD (0xBE). The patched version generates comma from shift period and behaves in a similar way to the other emulators.

**Comment by ZXFruitcake** [comment]
EightyOne appears to have been intended to map ZX81 symbols onto the keys of the Windows keyboard upon which they appear. Hence ‘;’ and ‘:’ are on the same key, ‘/’ and ‘?’ are on the same key, ',' and '<' are on the same key, etc. Yet none of these pair of symbols appear together on teh same key of a real ZX81 keyboard. Just modifying the key providing '.' and '>' would result in EightyOne following an inconsistent approach to handling keys of the Windows keyboard, which is not desirable.

An ideal solution would be to add new options on the Keyboard dialog to allow selection of what each symbol key on the Windows keyboard maps to. Alternatively, a faster solution to implement would be a checkbox to toggle between the current key mapping and a new mapping dedicated to the the USB keyboard, with the checkbox labelled based on the name of the USB keyboard (@ikjordan - does the keyboard have an official name yet?).

**Comment by ikjordan** [comment]
I think it is good that EightyOne allows the user to utilise the full Windows keyboard. However, for all of the other primary ZX81 keys on the Windows keyboard, the secondary function matches the ZX81. For example, Shift 2 generates AND not ", Shift 3 generates THEN not £, shift space generates £ etc, The . key seems the odd one out in that when shifted it generates the value expected by UK Windows, i.e. >.  

As you say, EightyOne has always done this, so I agree that my suggestion to change the default was wrong. Your proposal to add an option to the existing keyboard dialog box does seem a good one. 

I have updated my keyboard code so that it tracks the shift state and sends VK_OEM_COMMA rather than shift VK_OEM_PERIOD. This works with all the emulators I’ve tested, including EightyOne, so the trigger for my initial request has been resolved.
 
The code for the USB keyboard is at: https://github.com/ikjordan/ZX81_USB_KBD I’m afraid it does not have a very imaginative name :-)

Many thanks for maintaining EightyOne, it is a great piece of work. I’m currently running it, “typing” on a touch sensitive keyboard, and feeling like it is 1981 all over again!
